databuilder/databuilder/publisher/neo4j_csv_unwind_publisher.py

```diff
@@ -0,0 +1,380 @@
+
+    def _publish_relation_file(self, relation_file: str) -> None:
+        with open(relation_file, 'r', encoding='utf8') as relation_csv:
+            csv_dataframe = pandas.read_csv(relation_csv, na_filter=False)
+            all_rel_records = csv_dataframe.to_dict(orient="records")
+
+            # Get the first relation labels since they will be the same for all records in the file
+            merge_stmt = self.create_relationship_merge_statement(
+                rel_keys=csv_dataframe.columns.tolist(),
+                start_label=all_rel_records[0][RELATION_START_LABEL],
+                end_label=all_rel_records[0][RELATION_END_LABEL],
+                relation_type=all_rel_records[0][RELATION_TYPE],
+                relation_reverse_type=all_rel_records[0][RELATION_REVERSE_TYPE]
+            )
+
+            self._write_transactions(merge_stmt, all_rel_records)
+
+    def create_relationship_merge_statement(self,
+                                            rel_keys: list,
+                                            start_label: str,
+                                            end_label: str,
+                                            relation_type: str,
+                                            relation_reverse_type: str) -> str:
+        template = Template("""
+            UNWIND $batch as row
+            MATCH (n1:{{ START_LABEL }} {key: row.START_KEY}), (n2:{{ END_LABEL }} {key: row.END_KEY})
+            MERGE {{ relationship_stmt }}
+            {% if update_props_body %}
+            ON CREATE SET {{ props_body_create }}
+            ON MATCH SET {{ props_body_update }}
+            {% endif %}
+            RETURN n1.key, n2.key
+        """)
+
+        two_way_relationship = Template("""
+            (n1)-[r1:{{ TYPE }}]->(n2)-[r2:{{ REVERSE_TYPE }}]->(n1)
+        """).render(TYPE=relation_type, REVERSE_TYPE=relation_reverse_type)
+        one_way_relationship = Template("""
+            (n1)-[r1:{{ TYPE }}]->(n2)
+        """).render(TYPE=relation_type)
+
+        props_body_template = Template("""{{ props_body_r1 }} , {{ props_body_r2 }}""")
+
+        props_body_r1 = self._create_props_body(get_props_body_keys(rel_keys,
+                                                                    RELATION_REQUIRED_KEYS,
+                                                                    self._additional_publisher_metadata_fields), 'r1')
+        props_body_r2 = self._create_props_body(get_props_body_keys(rel_keys,
+                                                                    RELATION_REQUIRED_KEYS,
+                                                                    self._additional_publisher_metadata_fields), 'r2')
+        props_body_create = props_body_template.render(props_body_r1=props_body_r1, props_body_r2=props_body_r2)\
+            if self._publish_reverse_relationships else props_body_r1
+
+        props_body_update = props_body_create
+        if self._preserve_adhoc_ui_data:
```

Comment:
I left the ifs outside of the jinja template on the create node/rel stmt functions bc it saves from doing extra work if the preserve data config isn't set, and I felt like it would still be messy looking in any case, wdyt?